src/elexsolver/TransitionMatrixSolver.py

```diff
@@ -1,28 +1,107 @@
+        loss_function = cp.norm(A @ beta - B, "fro")
+        return cp.Minimize(loss_function)
+
+    def __ridge_objective(self, A: np.ndarray, B: np.ndarray, beta: np.ndarray) -> cp.Minimize:
+        # Based on https://www.cvxpy.org/examples/machine_learning/ridge_regression.html
+        lam = cp.Parameter(nonneg=True, value=self._lambda)
+        loss_function = cp.pnorm(A @ beta - B, p=2) ** 2
+        regularizer = cp.pnorm(beta, p=2) ** 2
+        return cp.Minimize(loss_function + lam * regularizer)
+
+    def __solve(self, A: np.ndarray, B: np.ndarray, weights: np.ndarray) -> np.ndarray:
+        transition_matrix = cp.Variable((A.shape[1], B.shape[1]), pos=True)
+        Aw = np.dot(weights, A)
+        Bw = np.dot(weights, B)
+
+        if self._lambda is None or self._lambda == 0:
+            objective = self.__standard_objective(Aw, Bw, transition_matrix)
+        else:
+            objective = self.__ridge_objective(Aw, Bw, transition_matrix)
+
+        constraints = TransitionMatrixSolver.__get_constraints(transition_matrix, self._strict)
+        problem = cp.Problem(objective, constraints)
+
+        with warnings.catch_warnings():
+            warnings.simplefilter("error")
+            try:
+                problem.solve(solver=cp.CLARABEL)
+            except (UserWarning, cp.error.SolverError) as e:
+                raise RuntimeError(e) from e
+
         return transition_matrix.value
 
-    def fit(self, A, B, strict=False):
-        transition_matrix = self.__solve(A, B, strict)
-        self.transition_matrix = transition_matrix
+    def fit_predict(self, X: np.ndarray, Y: np.ndarray, weights: np.ndarray | None = None) -> np.ndarray:
+        self._check_data_type(X)
+        self._check_data_type(Y)
+        self._check_any_element_nan_or_inf(X)
+        self._check_any_element_nan_or_inf(Y)
+
+        # matrices should be (units x things), where the number of units is > the number of things
+        if X.shape[1] > X.shape[0]:
+            X = X.T
+        if Y.shape[1] > Y.shape[0]:
+            Y = Y.T
+
+        if X.shape[0] != Y.shape[0]:
+            raise ValueError(f"Number of units in X ({X.shape[0]}) != number of units in Y ({Y.shape[0]}).")
+
+        self._check_dimensions(X)
+        self._check_dimensions(Y)
+        self._check_for_zero_units(X)
+        self._check_for_zero_units(Y)
+
+        if not isinstance(X, np.ndarray):
+            X = X.to_numpy()
+        if not isinstance(Y, np.ndarray):
+            Y = Y.to_numpy()
+
+        X_expected_totals = X.sum(axis=0) / X.sum(axis=0).sum()
+
+        X = self._rescale(X)
+        Y = self._rescale(Y)
+
+        weights = self._check_and_prepare_weights(X, Y, weights)
 
-    def predict(self, A):
-        return A @ self.transition_matrix
+        percentages = self.__solve(X, Y, weights)
+        self._transitions = np.diag(X_expected_totals) @ percentages
+        return percentages
```

Comment:
doesn't `fit_predict` usually return the labels? So in this case `self._transitions` rather than `percentages`? Is there any reason you didn't implement `fit` and `predict` and then have `fit_predict` call the two functions?

Reply:
Great questions that I will answer in reverse order:

1. I went with `fit_predict()` instead of `fit()` followed by `predict()` because in `scikit-learn` convention, one can `fit()` on one data set and then at some later point `predict()` on an entirely different data set (with the same features).  Here, the `predict()` has to happen on the same data set as the `fit()`, so I figured I'd keep with the `scikit-learn` convention for solvers where this is the case (e.g. clustering).  Also the `predict()` here would have to take nothing as an argument, which would break with `scikit-learn`'s `predict()` should we ever want to use anything from there e.g. grid search down the line. 🤷🏻‍♀️ 

2. This might be because my naming convention here is absolutely terrible so _please_ suggest new names 🙈  The transitions are the "flow-y bits" that get used in sankey diagrams.  If you sum the entire `self._transitions` matrix it should sum to 1 (100%).  However, the `percentages` are the beta terms that come out of the solver such that each row (representing a "from" candidate) sums to 1 (100%).  And only the `percentages` are being returned to again follow `scikit-learn` conventions where `fit_predict()` returns one thing, the predictions.

Reply:
Thanks for the detailed answers!
1. This solver is more general than the voterflow problem that we are using it for and there may be situations in which we (or someone else) will want to predict on data that was not used for fitting (for example if this were part of an election model, where we fit on a subset of the data but apply it to the full dataset).

2. The beta terms are not the predictions. The solution to our problem might be the coefficients of the optimization problem, but the predictions are still `y_hat = \hat{\beta} @ X`

Reply:
Oh, I see!  I'm so sorry, it never occurred to me that there might be multiple use-cases for this! 😓 

Ok, in that case I will:

* change this from `fit_predict()` to `fit()` and `predict()` where `fit()` returns nothing and `predict()` returns `Y_hat`;
* create another class member for the "percentages", probably just called `_betas` since that's more obvious lol;

Two questions though:

1. Does it make sense to keep calling this `Transition(Matrix)Solver`?
2. Does it make sense to do the math here for the `_transitions` or is that something exclusive to voterflow?  I do have spots in `elex-voterflow-model` where I do that math again (`_handle_non_voters()`) so if this is 💯 a voterflow thing, I can remove it from here and just keep it there.

🤔

Reply:
1. I'm not sure I follow why it wouldn't still be `TransitionMatrixSolver`?
2. Isn't computing transition what we'll be doing in the predict function?

Reply:
1. Yeah, thinking this through more, maybe you're right and the name works.  I was just thinking that, if `predict()` returns `Y_hat` instead of `betas`, which are what I've been calling the "percentages" and I think you're calling the "transitions", what is the goal of this solver?  Is it `Y_hat` or the `betas`?  But really, it's both, like with any other regression, really 🤔 
2. No, `predict()` will compute and return `Y_hat`.  This is awful terminology on my part and I'm so sorry 😬  You can't use the `betas` in the Sankey diagrams because they'll create a diagram that sums to more than 100% and displays the wrong per-candidate percentage totals on either side.  So we have to do math to create "transitions" or the flow-y pieces, as in: https://github.com/washingtonpost/elex-solver/blob/ELEX-3031-voterflow-models/src/elexsolver/TransitionMatrixSolver.py#L97 and https://github.com/WPMedia/elex-voterflow-model/blob/ELEX-3031-voterflow-models/src/elexvoterflow/voterflow.py#L198.  So what I'm wondering is, is this math to convert `betas` into the flow-y pieces something unique to the voterflow use-case?  If so, I'll remove it from the solver 🤔